src/content/docs/en/guides/incremental-static-regeneration.mdx

```diff
@@ -0,0 +1,199 @@
+---
+title: Incremental Static Regeneration
+---
+If your website fetches content from a CMS or a database that updates infrequently, you can use a technique called **Incremental Static Regeneration**, or ISR, to update your site without rebuilding the entire project. This is a great way to keep your site up-to-date while still enjoying the benefits of static hosting.
+
+:::note
+ISR is often presented as an alternative to SSG and SSR, but it's more helpful to think of it as SSR with caching.
+:::
+
+:::note
+Incremental static regeneration is different from **incremental builds**, which speeds up full rebuilds by reusing work done by the previous builds.
+:::
+
+## Why ISR?
+- Your site has a lot of pages (more than one thousand). Build times can be shortened when using SSR or ISR.
+- Your content changes infrequently (less frequently than once in 5 minutes on average).
+- Your site gets a lot of visitors. It can be wasteful to render the same page for every visitor, and you can avoid it using ISR.
+- The CMS or database you are using is slow, or charges by request. With ISR, the requests to your backend will become much less frequent.
+
+## Why not ISR?
+- Your content is located inside your project (for example, as markdown files in `src/content`).
+- You don't want the hassle of running a server or using a serverless platform.
+- Your website offers logged-in experiences. If each user needs to be sent a slightly different HTML from others, caching the page may not be possible.
+- The structure of your site (layouts, and styling) changes frequently. Changes to the structure require a full rebuild.
+
+## Implementing ISR for Your Project
+
+ISR involves saving the response such that it can be reused for future requests. The exact implementation depends on where you are deploying.
+
+### Self-hosted Node.js
+ISR can be implemented in Node.js by writing a caching middleware. The following middleware caches the response in memory for 60 seconds:
+
+```ts title="src/middleware.ts"
+
+const cache = {}
+
+export const onRequest = async ({ request }, next) => {
+    const cached = cache[request.url]
+    
+    if (cached) {
+        if (cached.expires > Date.now()) {
+            // return cached response
+            return cached.response.clone()
+        }
+        else {
+            // remove stale response
+            delete cache[request.url]
+        }
+    }
+    
+    else {
+        const response = await next()
+        cache[request.url] = {
+            expires: Date.now() + 60 * 1000,
+            response: response.clone()
+        }
+        // return fresh response
+        return response
+    }

```

Comment:
I mean like this, so when the request is cached but it is stale, after it being deleted the request is treated as non-cached. Otherwise, the function would return undefined when the cache is stale.

```suggestion
    const response = await next()
    cache[request.url] = {
        expires: Date.now() + 60 * 1000,
        response: response.clone()
    }
    // return fresh response
    return response

```

Reply:
You're right, my example did not work, thanks for this!

I would probably simplify it even more later. Fallthrough can be really tricky to think about, especially for new developers, which astro gets often.